fs_impl.go

```diff
@@ -0,0 +1,353 @@
+
+// Records contain sections saved as $fs_root/$uuid.$section The file payload
+// is saved in the 'data' section. metadata is in 'meta'. hooks can create own
+// sections
+
+import (
+	"encoding/json"
+	"errors"
+	"github.com/google/uuid"
+	"net/http"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+)
+
+const (
+	sectionPattern      = `[a-zA-Z0-9_-]+`
+	uuidPattern         = `[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}`
+	fileInFsRootPattern = uuidPattern + `(\.` + sectionPattern + `)?`
+
+	onlyUuidPattern         = `^` + uuidPattern + `$`
+	onlyFileInFsRootPattern = `^` + fileInFsRootPattern + `$`
+	onlySectionPattern      = `^` + sectionPattern + `$`
+)
+
+type Record struct {
+	Children []uuid.UUID `json:"children,omitempty"`
+	IsDir    bool        `json:"is_dir"`
+	Name     string      `json:"name"`
+	refs     uint        `json:"-"`
+}
+
+type Fs struct {
+	records  map[uuid.UUID]*Record
+	root     uuid.UUID
+	basePath string
+}
+
+func (fs *Fs) path(p string) string {
+	// TODO sanitize paths
+	return filepath.Join(fs.basePath, p)
+}
+
+func (fs *Fs) writeToDisk(id uuid.UUID, r *Record) error {
+	f, err := os.Create(fs.path(id.String()))
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	enc := json.NewEncoder(f)
+	return enc.Encode(r)
+}
+
+func (fs *Fs) newRecord(parent_uuid uuid.UUID, name string) (uuid.UUID, error) {
+	new_id := uuid.New()
+
+	new_rec := new(Record)
+	new_rec.Children = []uuid.UUID{}
+	new_rec.Name = name
+	new_rec.refs = 1
+
+	fs.records[new_id] = new_rec
+
+	parent_record := fs.records[parent_uuid]
+	parent_record.Children = append(parent_record.Children, new_id)
+	fs.records[parent_uuid] = parent_record
+
+	return new_id, fs.writeToDisk(new_id, new_rec)
+}
+
+// return new slice that does not contain v
+func removeUUID(s []uuid.UUID, v uuid.UUID) []uuid.UUID {
+	i := 0
+	pos := -1
+	for ; i < len(s); i++ {
+		if s[i] == v {
+			pos = i
+			break
+		}
+	}
+
+	for ; i < len(s); i++ {
+		if s[i] == v {
+			panic("duplicite uuid")
+		}
+	}
+
+	if pos == -1 {
+		panic("uuid not found")
+	}
+
+	// swap remove
+	s[pos] = s[len(s)-1]
+	return s[:len(s)-1]
+}
+
+func (fs *Fs) deleteRecordFilesFromDisk(name string) error {
+	entries, err := os.ReadDir(fs.basePath)
+	if err != nil {
+		return err
+	}
+
+	// list of all errors. We try to delete as many files as we can to
+	// avoid zombie files and then return one big error with all of the
+	// failed removes.
+	var errs []error
+
+	err = os.Remove(fs.path(name))
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	// TODO(prokop) keep a list of all sections instead of iterating over
+	// all files
+	for _, e := range entries {
+		if strings.HasPrefix(e.Name(), name+".") {
+			err = os.Remove(fs.path(e.Name()))
+			if err != nil {
+				errs = append(errs, err)
+			}
+		}
+	}
+
+	if len(errs) != 0 {
+		return errors.Join(errs...)
+	}
+	return nil
+}
+
+func (fs *Fs) deleteChildInMemoryAndWriteParentToDisk(parent uuid.UUID, child uuid.UUID) error {
```

Comment:
Why not just `deleteChild`? Why does the function name have to mention that the child is deleted from memory and the parent is saved to the disk? This is obvious behavior, which should at max be mentioned in a comment above the function.